File: client/src/boundary/interactor/ShotResultD2BConverter.ts
import { Converter } from "utils/Converter";
import { ShotResult } from "domain/index";
import { BoundaryShotResult } from "boundary/model/index";
import { ShipD2BConverter } from "./ShipD2BConverter";

export class ShotResultD2BConverter extends Converter<
  ShotResult,
  BoundaryShotResult
> {
  private shipConverter: ShipD2BConverter;

  constructor(shipConverter: ShipD2BConverter) {
    super();
    this.shipConverter = shipConverter;
  }

  convert(shotResult: ShotResult): BoundaryShotResult {
    return new BoundaryShotResult(
      shotResult.grid,
      shotResult.gameState,
      shotResult.ship ? this.shipConverter.convert(shotResult.ship) : undefined
    );
  }
}

File: server/battleships/use_cases/implementation/src/main/java/eu/optas/use_cases/implementation/ShotResultD2BConverter.java
package eu.optas.use_cases.implementation;

import eu.optas.domain.Ship;
import eu.optas.domain.ShotResult;
import eu.optas.use_cases.api.BoundaryShip;
import eu.optas.use_cases.api.BoundaryShotResult;
import eu.optas.utils.Converter;

public class ShotResultD2BConverter extends Converter<ShotResult, BoundaryShotResult> {

    @Override
    public BoundaryShotResult convert(ShotResult shotResult) {
        BoundaryShip boundaryShip = shotResult.getShip()
                .map(this::convertShip)
                .orElse(null);

        return new BoundaryShotResult(shotResult.getGrid(), shotResult.getGameState(), boundaryShip);
    }

    private BoundaryShip convertShip(Ship ship) {
        return new BoundaryShip(ship.getLength(), ship.getCoordinates(), ship.getHits(), ship.isDestroyed());
    }
}
